rfd/0xxx-partial-cache-healthiness.md

```diff
@@ -0,0 +1,332 @@
+be attached to a new special resource type associated with the event. For backwards compatibility, the watch request
+should fail if the new flag wasn't set on it.
+
+Cache will use the rejected kinds list for controlling initial cache synchronization, read request routing and
+propagation of the rejected status to fanout event readers.
+
+During the initial synchronization, cache will delete all configured resource kinds from the in-memory storage, including
+the rejected ones, to avoid potential reads from stale data. Then, only non-rejected kinds will be fetched and upserted
+into the storage.
+
+The routing logic implemented in the `readGuard` component will also need to be updated. Currently, it's binary:
+if the cache is healthy, all reads are routed to the local services powered by the in-memory backend,
+or to the API client otherwise. Read guard should become aware of the resource kinds that were rejected by the event source
+and route requests related to those resource kinds to the API client even when cache is in a healthy state.
+
+Throughout its lifecycle, cache can encounter disconnections and other errors interrupting the event stream. This brings
+it back to unhealthy state. Every time it becomes healthy again, it's considered a new *generation* of cache. Since every
+time it reconnects to the Auth service, it might be of different versions, the list of rejected kinds should be considered
+specific to the current generation and refreshed during each re-initialization.
+
+Since cache can also act as an event source, it should propagate the rejected status if a client tries to use the cache
+to subscribe to events of an object kind that was rejected to the cache itself. Such client will also receive a
+list of rejected kinds on the first `OpInit` event or a failure depending on the supplied AllowRejectedKinds flag. 
+
+
+### Changes to the Events interface
+
+Because the `Events` interface is central to the caching subsystem, it will require some changes. Also, because it has
+so many implementations and is used in so many contexts, some rules need to be established and all the implementations
+must be updated to behave in a consistent way.
+
+None of the proposed changes to the interface are direct though. They all happen through related types and behaviour.
+
+
+#### types.Watch
+```go
+// Watch sets up watch on the event
+type Watch struct {
+	Name string
+	Kinds []WatchKind
+	QueueSize int
+	MetricComponent string
+
+	// *NEW* AllowRejectedKinds enables partial success mode in which the request won't fail if some of the requested Kinds
+	// are not available for watching.
+	AllowRejectedKinds bool
+
+	// *NEW* PreviouslyRejectedKinds contains a list of kinds rejected by upstream implementations in a chain of NewWatcher()
+	// calls. Must be empty if AllowRejectedKinds is not set. A kind cannot be on both Kinds and PreviouslyRejectedKinds
+	// at the same time.
+	PreviouslyRejectedKinds []RejectedKind
```

Comment:
We could also sneak this information into inner implementations as a value on `ctx` like we do with loggers and other things. I'm not a fan of that approach because hiding things away can lead to future programming errors, but it can make this API look tidier on the outside.